Cannot find module `pages/_document`

# Bug report

## Describe the bug

Upgrading to 7 from 6 and application works fine in development mode. In production, all pages fail with the server error `Cannot find module '/app/.next/server/static/ddabb420240ac61ad07cc315c21cfa96e55c20df/pages/_document'`.

Not using a custom document, but am using a custom `_app.js`. Verified that `_document.js` exist at that path in `.next`.

## To Reproduce

Upgrade to Next 7.

## Expected behavior

Pages should load without the error.

## System information
`package.json`:
```
"dependencies": {
		"@babel/core": "7.1.0",
		"@babel/register": "7.0.0",
		"@fortawesome/fontawesome-svg-core": "1.2.4",
		"@fortawesome/free-brands-svg-icons": "5.3.1",
		"@fortawesome/free-regular-svg-icons": "5.3.1",
		"@fortawesome/free-solid-svg-icons": "5.3.1",
		"@fortawesome/react-fontawesome": "0.1.3",
		"@zeit/next-css": "1.0.1",
		"aws-sdk": "2.317.0",
		"babel-plugin-inline-import": "3.0.0",
		"babel-plugin-inline-svg": "1.0.0",
		"clipboard": "2.0.1",
		"css-loader": "1.0.0",
		"csv-parse": "3.0.0",
		"enum": "2.5.0",
		"file-loader": "1.1.11",
		"js-md5": "0.7.3",
		"luxon": "1.3.3",
		"newrelic": "4.8.1",
		"next": "7.0.0",
		"path-match": "1.2.4",
		"promise-file-reader": "1.0.2",
		"react": "16.5.2",
		"react-autosuggest": "9.4.1",
		"react-dom": "16.5.2",
		"react-load-script": "0.0.6",
		"react-phone-number-input": "2.2.5",
		"react-places-autocomplete": "7.2.0",
		"react-tagsinput": "3.19.0",
		"socket.io-client": "2.1.1",
		"sw-loader": "0.1.2",
		"timeago.js": "3.0.2",
		"uuid": "3.3.2"
	}
```

As @a-kon mentioned, currently any errors that occur when building `_document` are swallowed, making issues very hard to debug. In my case, I didn't immediately notice that I was doing:

```js
import NextDocument, { Head, Main, NextScript } from "next/document";
import Head from "next/head";
```

as the error was simply:

```
Error: Cannot find module '…/.next/server/static/development/pages/_document'
```

--- a/errors/generatebuildid-not-a-string.md
+++ b/errors/generatebuildid-not-a-string.md
@@ -1,0 +1,9 @@
+# generateBuildId did not return a string
+
+#### Why This Error Occurred
+
+The most common cause for this issue is a custom `next.config.js` with the `generateBuildId` method defined, but it does not return a string.
+
+#### Possible Ways to Fix It
+
+Always return a string from generateBuildId.

--- a/packages/next/build/index.js
+++ b/packages/next/build/index.js
@@ -37,7 +37,7 @@ export default async function build (dir, conf = null) {
   const distDir = join(dir, config.distDir)
   const buildId = await generateBuildId(config.generateBuildId, nanoid)
 
-  await ensureProjectDirectoryIsWriteAble()
+  await ensureProjectDirectoryIsWriteAble(dir)
 
   try {
     const configs = await Promise.all([

--- a/packages/next/build/index.js
+++ b/packages/next/build/index.js
@@ -10,22 +10,34 @@ import getBaseWebpackConfig from './webpack'
 const access = promisify(fs.access)
 const writeFile = promisify(fs.writeFile)
 
-export default async function build (dir, conf = null) {
-  const config = loadConfig(PHASE_PRODUCTION_BUILD, dir, conf)
-  const distDir = join(dir, config.distDir)
-
-  let buildId = await config.generateBuildId() // defaults to a uuid
-  if (buildId == null) {
-    // nanoid is a small url-safe uuid generator
-    buildId = nanoid()
+async function generateBuildId (generate, fallback) {
+  let buildId = await generate()
+  // If there's no buildId defined we'll fall back
+  if (buildId === null) {
+    buildId = fallback()
   }
 
+  if (typeof buildId !== 'string') {
+    throw new Error('generateBuildId did not return a string. https://err.sh/zeit/next.js/generatebuildid-not-a-string')
+  }
+
+  return buildId.trim()
+}
+
+async function ensureProjectDirectoryIsWriteAble (dir) {
   try {
     await access(dir, (fs.constants || fs).W_OK)
   } catch (err) {
-    console.error(`> Failed, build directory is not writeable. https://err.sh/zeit/next.js/build-dir-not-writeable`)
-    throw err
+    throw new Error('Build directory is not writeable. https://err.sh/zeit/next.js/build-dir-not-writeable')
   }
+}
+
+export default async function build (dir, conf = null) {
+  const config = loadConfig(PHASE_PRODUCTION_BUILD, dir, conf)
+  const distDir = join(dir, config.distDir)
+  const buildId = await generateBuildId(config.generateBuildId, nanoid)
+
+  await ensureProjectDirectoryIsWriteAble()
 
   try {
     const configs = await Promise.all([

